src/HexagonOptimize.cpp

```diff
@@ -341,16 +341,44 @@ class OptimizePatterns : public IRMutator {
     }
 
     void visit(const Add *op) {
+        // vmpa instructions have a lot of quirky conditions and
+        // matching operations, and would require flags used only by
+        // these instructions (a double narrowing). Instead, just
+        // handle them manually here.
+        Expr vmpa_ub_pattern = wild_i16x*bc(wild_i16) + wild_i16x*bc(wild_i16);
+        Expr vmpa_h_pattern = wild_i32x*bc(wild_i32) + wild_i32x*bc(wild_i32);
+        vector<Expr> matches;
+        if (expr_match(vmpa_ub_pattern, op, matches)) {
+            // Narrow the operands.
+            Expr v0 = lossless_cast(UInt(8).with_lanes(op->type.lanes()), matches[0]);
+            Expr v1 = lossless_cast(UInt(8).with_lanes(op->type.lanes()), matches[2]);
+            Expr c0 = lossless_cast(Int(8), matches[1]);
+            Expr c1 = lossless_cast(Int(8), matches[3]);
+
+            if (v0.defined() && v1.defined() && c0.defined() && c1.defined()) {
+                expr = halide_hexagon_add_mpy_mpy(v0, v1, c0, c1, ".vub.vub.b.b");
```

Comment:
I am thinking out aloud here so please bear with me. Do we expect to match these patterns only at the leaves? If so, this is fine, but if not, then shouldn't we be stepping into the v0 and v1 by mutating them.

I guess the lossless_cast means that we'll mostly be matching casts of vectors and it is perhaps rare to see something like

cast<int16_t>((v0 + v1)) * 3 + 4 * cast<int16_t>(v2)